docs/adr/adr1-types.md

```diff
@@ -0,0 +1,352 @@
+### 2.2. Type annotations as assumptions
+
+One can use TLA+ syntax to write assumptions and assertions about the types.
+We are talking only about type assumptions here.
+The similar approach can be used to write theorems about types. 
+Consider the following specification:
+
+```tla
+EXTENDS Sequences
+CONSTANTS Range
+VARIABLES list
+
+Mem(e, es) ==
+  \E i \in DOMAIN es:
+    e = es[i]
+```
+
+In this example, the operator `Mem` is polymorphic, whereas the types of `Range`
+and `list` are parameterized.  If the user wants to
+restrict the types of constants, variables, and operators, they could write (using the
+[TypeOK syntax](#typesAsTypeOk)):
+
+```tla
+ASSUME(Range \in SUBSET Int)
+ASSUME(list \in Seq(Int))
+ASSUME(\A e \in Int, \A es \in Seq(Int): Mem(e, es) \in BOOLEAN)
+```
+
+Moreover, using the proof syntax of
+[TLA+ Version 2](https://lamport.azurewebsites.net/tla/tla2-guide.pdf),
+we can annotate the
+types of variables introduced inside the operators.  For instance, we could
+label the name `i` as follows:
+
+```tla
+Mem(e, es) ==
+  \E i \in DOMAIN es:
+    e = es[i_use :: i]
+```
+
+And then write:
+
+```tla
+ASSUME(\A e, es, i: Mem(e, es)!i_use(i) \in Int)
+```
+
+__Pros__:
+
+ * The assumptions syntax is quite appealing, when writing types of
+    CONSTANTS, VARIABLES, and top-level operators.
+
+__Cons__:
+
+ * The syntax gets verbose and hard to write, when writing types of
+   LET-IN operators and bound variables.
+ * It is not clear how to extend this syntax to higher-order operators.
+
+### 2.3. Type annotations in comments
+ 
+This solution basically gives up on TLA+ syntax and introduces a special
+syntax à la javadoc for type annotations:
+
+```tla
+EXTENDS Sequences
+CONSTANTS Range \*@ Range: Set(Int)
+VARIABLES list  \*@ list: Seq(Int)
+
+Mem(e, es) ==
+\*@ Mem: (Int, Seq(Int)) => Bool
+  \E i \in DOMAIN es:
+    \*@ i: Int
+    e = es[i]
+```
+
+We have not come up with a good syntax for these annotaions. The above
+example demonstrates one possible approach.
+
+__Pros__:
+
+  * Non-verbose, simple syntax
+  * Type annotations do not stand in the way of the specification author
+  * Type annotations may be collapsed, removed, etc.
+  * If we have an annotation preprocessor, we can use it for other
+    kinds of annotations
+
+__Cons__:
+
+  * As we give up on the TLA+ syntax, TLA+ Toolbox will not help us
+    (though it is not uncommon for IDEs to parse javadoc annotations,
+     so there is some hope)
+  * The users have to learn new syntax for writing type annotations and types
+  * We have to write an annotation preprocessor
+
+### 2.4. Type annotations as operator definitions
+
+Operators definitions and LET-IN definitions can be written almost anywhere in
+TLA+. Instead of writing in-comment annotations, we can write annotations
+with operator definitions (assuming [types as strings](#typesAsStrings),
+but this is not necessary):
+
+```tla
+EXTENDS Sequences
+CONSTANTS Range
+Range_type == "set(z)"
```

Comment:
Aside from aesthetic concerns, one thing i worry about here is that it seems all to easy for users to make a typo that would cuase an intended type annotation to end up being an unused operator definition. e.g.,

```
Range_typ == "set(z)"
```

This would just get ignored, right? Because it's just defining a nullary operator. So it is then up to user diligence. I think all of the other solutions would protect against this kind of thing. Perhaps this could be addressed with linting that checks for unused operators tho..

Reply:
Yep, a typo would ruin the annotation.